### PR TITLE
[develop] Trello-zgEO6IAA: Migrator and migration for creating currencies

### PR DESCRIPTION
**Pull request by AndreySurzhan**
https://trello.com/c/zgEO6IAA

- [x] Implement migrator.js
- [x] Implement migration for adding currencies
- [x] create and update models for currencies and migrations
- [x] code review

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,4 @@
 .idea/
-node_modules/
\ No newline at end of file
+node_modules/
+.DS_Store
+models/.DS_Store
```
Comment (wildeastengineer) [inline .gitignore]: add empty line, please

File: migrations/201606131415_currenciesListCreation.js [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        },
+        {
+            iso: 'JPY'
+        }
+    ];
+    var deferred = Q.defer();
+
+    currencyCollection = db.collection(collectionName);
+
+    currencyCollection.find({}, {iso: 1, _id: 0}).toArray().then(function (docs) {
```
Comment (wildeastengineer) [inline migrations/201606131415_currenciesListCreation.js]: object's fields on new line

File: migrations/201606131415_currenciesListCreation.js [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        var currenciesForAdding = [];
+        var i;
+        var j;
+
+        if (docs.length === 0) {
+            logger.info('Collection "{collection}" is empty - all docs should be added'
```
Comment (wildeastengineer) [inline migrations/201606131415_currenciesListCreation.js]: Lets use "${collectionName}". After ES6 support implementing we will be able just to remove `.replace('${collectionName}', collectionName)` in this case

File: migrations/201606131415_currenciesListCreation.js [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                    .replace('{collection}', collectionName));
+            currenciesForAdding = currenciesList;
+        } else {
+            for (i = 0; i < currenciesList.length; i++) {
+                for (j = 0; j < docs.length; j++) {
+                    if (currenciesList[i].iso === docs[j].iso) {
+                        break;
+                    }
+
+                    if (j + 1 === docs.length) {
```
Comment (wildeastengineer) [inline migrations/201606131415_currenciesListCreation.js]: What does it mean?

File: migrations/201606131415_currenciesListCreation.js [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+                        currenciesForAdding.push(currenciesList[i]);
+                    }
+                }
+            }
+        }
+
+        if (currenciesForAdding.length === 0) {
+            logger.info('Collection "{collection}" already has all necessary docs'
```
Comment (wildeastengineer) [inline migrations/201606131415_currenciesListCreation.js]: `${collectionName}`

File: migrations/201606131415_currenciesListCreation.js [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+            return deferred.promise;
+        }
+
+        currencyCollection.insertMany(currenciesForAdding).then(function (result) {
+            deferred.resolve();
+        }).catch(function (err) {
+            deferred.reject(err);
+        });
+
```
Comment (wildeastengineer) [inline migrations/201606131415_currenciesListCreation.js]: extra line

File: migrations/201606131415_currenciesListCreation.js [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        if (currenciesForAdding.length === 0) {
+            logger.info('Collection "{collection}" already has all necessary docs'
+                    .replace('{collection}', collectionName));
+            deferred.resolve();
+
+            return deferred.promise;
+        }
+
+        currencyCollection.insertMany(currenciesForAdding).then(function (result) {
```
Comment (wildeastengineer) [inline migrations/201606131415_currenciesListCreation.js]: What do you think about placing `then` on new line?

```
currencyCollection.insertMany(currenciesForAdding)
    .then(function (result) {
        deferred.resolve();
    })
    .catch(function (err) {
        deferred.reject(err);
    });
```

File: migrations/migrator.js [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    /**
+     * Executes migrations and write information about execution to database
+     *
+     * @param {string} dbUrl
+     *
+     * @returns {promise}
+     */
+    run: function (dbUrl) {
+        var deferred = Q.defer();
+        var MongoClient = require('mongodb').MongoClient;
```
Comment (wildeastengineer) [inline migrations/migrator.js]: on the top of page

File: migrations/migrator.js [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+            async.each(getAllMigrations(), function (migration, callback) {
+                findAndExecuteMigration(migration).then(function () {
+                    callback();
+                }).catch(function (err) {
+                    callback(err);
+                });
+            }, function (err) {
+                if (err) {
+                    db.close();
```
Comment (wildeastengineer) [inline migrations/migrator.js]: put `db.close();` before `if`

File: migrations/migrator.js [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+             * @private
+             * @param {Object[]} migration
+             *
+             * @returns {promise}
+             */
+            function findAndExecuteMigration (migration) {
+                var deferred = Q.defer();
+                var executeMigration = require('./' + migration.fileName);
+
+                migrationCollection.find(
```
Comment (wildeastengineer) [inline migrations/migrator.js]: migrationCollection.find({
    name: migration.name
})

File: migrations/migrator.js [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+                    {
+                        name: migration.name
+                    }
+                ).toArray().then(function (docs) {
+                    var i;
+
+                    if (docs.length > 0) {
```
Comment (wildeastengineer) [inline migrations/migrator.js]: There is something wrong..

If we want to allow being 2 migrations with equal names but with different dates, then we should fix error message in logger calling: inform about date, not about existing.

If we want to forbid equal names, the we should terminate migration execution without date checking

File: package.json [hunk 12]
```diff
@@ -2,10 +2,12 @@
   "name": "mk-api",
   "main": "server.js",
   "dependencies": {
+    "async": "^2.0.0-rc.6",
     "body-parser": "1.0.1",
     "crypto": "0.0.3",
+    "database": "3.6.13",
     "express": "4.0.0",
-    "mongoose": "3.6.13",
```
Comment (wildeastengineer) [inline package.json]: `mongoose` should not be deleted

File: models/migration.js [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+        type: String,
+        required: true
+    },
+    fileName: {
+        type: String,
+        unique: true,
+        required: true
+    },
+    date: {
+        type: String,
```
Comment (AndreySurzhan) [inline models/migration.js]: should be unique

Comment (AndreySurzhan) [inline models/migration.js]: done

File: .gitignore (hunk 1 above)
Comment (AndreySurzhan) [inline .gitignore]: done

File: migrations/201606131415_currenciesListCreation.js [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+                    });
+                }
+            } else {
+                for (i = 0; i < currenciesList.length; i++) {
+                    for (j = 0; j < docs.length; j++) {
+                        if (currenciesList[i] === docs[j].iso) {
+                            break;
+                        }
+
+                        if (j === docs.length - 1) {
```
Comment (AndreySurzhan) [inline migrations/201606131415_currenciesListCreation.js]: this checks if there was not currency from docs in currenciesList till loop is ended then add missing currency into currenciesForAdding

**Comment by AndreySurzhan** [comment]
@wildeastengineer review please

File: migrations/201606131415_currenciesListCreation.js [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+                iso: 1,
+                _id: 0
+            }).toArray();
+        })
+        .then(function (docs) {
+            var currenciesForAdding = [];
+            var i;
+            var j;
+
+            if (docs.length === 0) {
```
Comment (wildeastengineer) [inline migrations/201606131415_currenciesListCreation.js]: I think we can omit this and do only `else` branch

File: migrations/201606131415_currenciesListCreation.js [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+                for (i = 0; i < currenciesList.length; i++) {
+                    currenciesForAdding.push({
+                        iso: currenciesList[i],
+                        created: new Date(),
+                        modified: new Date()
+                    });
+                }
+            } else {
+                for (i = 0; i < currenciesList.length; i++) {
+                    for (j = 0; j < docs.length; j++) {
```
Comment (wildeastengineer) [inline migrations/201606131415_currenciesListCreation.js]: Maybe it's better to use [underscore](https://github.com/jashkenas/underscore) here to find currency in a list

File: migrations/migrator.js [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+
+            /**
+             * @private
+             * @param {string} migrationFile
+             *
+             * @returns {promise}
+             */
+            function findAndExecuteMigration (migrationFile) {
+                var deferred = Q.defer();
+                var executeMigration = require('./' + migrationFile);
```
Comment (wildeastengineer) [inline migrations/migrator.js]: I think it as not really correct. To this variable you should get _migration_, not _**execute**Migration_.
And migrations's module should has method `run` or `execute`. It seems to be more readable.

File: migrations/migrator.js [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+                var migration = {
+                    name: getMigrationName(migrationFile),
+                    date: getMigrationDate(migrationFile),
+                    fileName: migrationFile
+                };
+
+                migrationCollection.find({}).toArray().then(function (docs) {
+                    var i;
+
+                    if (docs.length > 0) {
```
Comment (wildeastengineer) [inline migrations/migrator.js]: It's better to extract searching code to separate method

**Comment by AndreySurzhan** [comment]
@wildeastengineer review please

File: migrations/migrator.js [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+    run: function (dbUrl) {
+        var deferred = Q.defer();
+
+        MongoClient.connect(dbUrl, function(err, db) {
+            var migrationCollection;
+
+            migrationCollection = db.collection('migrations');
+            
+            migrationCollection.ensureIndex({
+                    date: 1,
```
Comment (wildeastengineer) [inline migrations/migrator.js]: What is the meaning of date uniq?

File: migrations/migrator.js (hunk 17 above)
Comment (wildeastengineer) [inline migrations/migrator.js]:
```
var migration = require(...
```

File: migrations/migrator.js [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+                 *
+                 * @returns {boolean}
+                 */
+                function isExecuted (migration, executedMigrations) {
+                    var i;
+
+                    if (executedMigrations.length > 0) {
+                        for (i = 0; i < executedMigrations.length; i++) {
+                            if (migration.date <= executedMigrations[i].date) {
+                                logger.info('Skip migration "${migration}". It already exists or it is not up-to-date '
```
Comment (wildeastengineer) [inline migrations/migrator.js]: I think we should skip migration if it is already executed and terminate migrations executing if it is out of date

File: models/migration.js [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+        unique: true,
+        required: true
+    },
+    created: {
```
Comment (wildeastengineer) [inline models/migration.js]: what is `created`?

File: migrations/migrator.js (hunk 19 above)
Comment (AndreySurzhan) [inline migrations/migrator.js]: @wildeastengineer you mean that fileName uniqueness would be enough, right?

File: models/migration.js (hunk 21 above)
Comment (AndreySurzhan) [inline models/migration.js]: I thought we have it as a standard already.
Each entity in DB has ```created``` field that means actual date of creation that entity in data base

```date``` filed can be very different from ``created``` date

File: migrations/migrator.js (hunk 20 above)
Comment (AndreySurzhan) [inline migrations/migrator.js]: @wildeastengineer not sure whether I understand what you mean.
We check that migration by date that current one is older than all existing ones

**Comment by AndreySurzhan** [comment]
@wildeastengineer right now the logic of migrator is following:
check through migrations list
- if migration is older then all migrations in DB then fail migrating on that step
- if it has the same date with migration that is in DB then just skip and and continue migrating
- if migration is up-to-date then execute it

**Comment by AndreySurzhan** [comment]
@wildeastengineer pelase review

**Comment by wildeastengineer** [comment]
:+1:

**Pull request merged**